File: ruby/lib/epostak/resources/inbound.rb
# frozen_string_literal: true

require "erb"

module EPostak
  module Resources
    # Resource for the Pull API inbound (received) documents.
    #
    # This is the preferred path for reading received invoices. Unlike the
    # legacy +/documents/inbox+ endpoint (enterprise-only), this resource
    # requires only the +api-eligible+ plan.
    #
    # Accessed via +client.inbound+.
    #
    # @example List and acknowledge received documents
    #   page = client.inbound.list(limit: 50)
    #   page["documents"].each do |doc|
    #     process(doc)
    #     client.inbound.ack(doc["id"])
    #   end
    class Inbound
      # @param http [EPostak::HttpClient] Internal HTTP client
      def initialize(http)
        @http = http
      end

      # List received inbound documents (cursor-paginated).
      #
      # @param since [String, nil] ISO 8601 lower bound (exclusive)
      # @param limit [Integer] Max documents to return (1–500, default 100)
      # @param kind [String, nil] Filter by document kind (e.g. "invoice", "credit_note", "self_billing")
      # @param sender [String, nil] Filter by sender Peppol ID
      # @return [Hash] { "documents" => [...], "next_cursor" => String|nil, "has_more" => bool }
      #
      # @example
      #   page = client.inbound.list(since: "2026-05-01T00:00:00Z", limit: 100)
      #   page["documents"].each { |d| puts d["id"] }
      def list(since: nil, limit: 100, kind: nil, sender: nil)
        query = { since: since, limit: limit, kind: kind, sender: sender }
        @http.request(:get, "/inbound/documents", query: query)
      end

      # Fetch a single inbound document by ID.
      #
      # @param id [String] Document UUID
      # @return [Hash] Full inbound document object
      #
      # @example
      #   doc = client.inbound.get("doc-uuid")
      #   puts doc["senderPeppolId"]
      def get(id)
        @http.request(:get, "/inbound/documents/#{encode(id)}")
      end

      # Download the raw UBL XML for an inbound document.
      #
      # @param id [String] Document UUID
      # @return [String] UBL 2.1 XML as a raw string
      # @raise [EPostak::Error] 404 when no XML is available (legacy rows without rawXmlPath)
      #
      # @example
      #   xml = client.inbound.get_ubl("doc-uuid")
      #   File.write("received.xml", xml)
      def get_ubl(id)
        @http.request_raw(:get, "/inbound/documents/#{encode(id)}/ubl")
      end

      # Acknowledge a received inbound document (marks it as processed on your side).
      #
      # Idempotent — re-acknowledging the same document overwrites +clientAckedAt+
      # with the latest timestamp.
      #
      # @param id [String] Document UUID
      # @param client_reference [String, nil] Optional client-side reference (max 256 chars)
      # @return [Hash] Updated inbound document shape
      #
      # @example
      #   result = client.inbound.ack("doc-uuid", client_reference: "PROC-2026-0042")
      #   puts result["clientAckedAt"]
      def ack(id, client_reference: nil)
        body = {}
        body[:client_reference] = client_reference if client_reference
        @http.request(:post, "/inbound/documents/#{encode(id)}/ack", body: body)
      end

      private

      def encode(value)
        ERB::Util.url_encode(value)
      end
    end
  end
end

File: python/src/epostak/resources/inbound.py
"""Inbound resource -- Pull API for received documents.

Provides :class:`InboundResource` accessible via ``client.inbound``.
Requires ``api-eligible`` plan (``requireApiEligiblePlan``), scope ``documents:read``
for list/get/get_ubl and ``documents:write`` for ack.
"""

from __future__ import annotations

from typing import Any, Dict, Optional, TYPE_CHECKING
from urllib.parse import quote

if TYPE_CHECKING:
    from epostak.types import (
        PullInboundDocument,
        PullInboundListResponse,
        PullInboundAckResponse,
    )

from epostak.resources.documents import _BaseResource, _build_query, _idempotency_headers


class InboundResource(_BaseResource):
    """Pull API: received (inbound) documents."""

    def list(
        self,
        *,
        since: Optional[str] = None,
        limit: Optional[int] = None,
        kind: Optional[str] = None,
        sender: Optional[str] = None,
    ) -> PullInboundListResponse:
        """Fetch received documents (cursor-paginated).

        Args:
            since: ISO 8601 cursor — only return documents received after this timestamp.
            limit: Max documents to return, 1–500 (default 100).
            kind: Filter by document kind, e.g. ``"invoice"``, ``"credit_note"``,
                ``"self_billing"``.
            sender: Filter by sender Peppol ID, e.g. ``"0245:1234567890"``.

        Returns:
            Dict with ``documents``, ``next_cursor`` (str or None), and
            ``has_more`` (bool).

        Example::

            page = client.inbound.list(limit=50)
            for doc in page["documents"]:
                print(doc["id"], doc["sender_peppol_id"])
            if page["has_more"]:
                next_page = client.inbound.list(since=page["next_cursor"])
        """
        params = _build_query({"since": since, "limit": limit, "kind": kind, "sender": sender})
        return self._request("GET", "/inbound/documents", params=params)

    def get(self, id: str) -> PullInboundDocument:
        """Get a single received document by ID.

        Args:
            id: Document UUID.

        Returns:
            Full inbound document object.

        Example::

            doc = client.inbound.get("doc-uuid")
            print(doc["status"], doc["sender_peppol_id"])
        """
        return self._request("GET", f"/inbound/documents/{quote(id, safe='')}")

    def get_ubl(self, id: str) -> str:
        """Get the raw UBL XML for a received document.

        Returns ``application/xml`` from the server. Raises ``EPostakError``
        with status 404 when the document has no stored raw XML (legacy rows
        created before UBL storage was enabled).

        Args:
            id: Document UUID.

        Returns:
            UBL 2.1 XML string.

        Example::

            xml = client.inbound.get_ubl("doc-uuid")
            with open("invoice.xml", "w", encoding="utf-8") as f:
                f.write(xml)
        """
        resp = self._request("GET", f"/inbound/documents/{quote(id, safe='')}/ubl", raw=True)
        if isinstance(resp, str):
            return resp
        # httpx Response object from raw=True
        try:
            return resp.text
        except AttributeError:
            return str(resp)

    def ack(
        self,
        id: str,
        *,
        client_reference: Optional[str] = None,
        idempotency_key: Optional[str] = None,
    ) -> PullInboundAckResponse:
        """Acknowledge a received document (idempotent, latest-ack-wins).

        Marks the document as client-acknowledged. Subsequent acks overwrite
        ``clientAckedAt`` (latest-ack-wins). Scope required: ``documents:write``.

        Args:
            id: Document UUID to acknowledge.
            client_reference: Optional client-side reference string (max 256 chars).
                Stored on the document and returned in the response.
            idempotency_key: Optional idempotency key to prevent duplicate acks.

        Returns:
            Full document object post-ack.

        Example::

            doc = client.inbound.ack("doc-uuid", client_reference="PO-2026-001")
            print(doc["client_acked_at"])
        """
        body: Dict[str, Any] = {}
        if client_reference is not None:
            body["client_reference"] = client_reference
        return self._request(
            "POST",
            f"/inbound/documents/{quote(id, safe='')}/ack",
            json=body,
            extra_headers=_idempotency_headers(idempotency_key),
        )
